File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/src/LIVE_short_trader_multi_filter/order_utils.py
from __future__ import annotations

import random
import time
from dataclasses import dataclass
from typing import Optional, Sequence, Tuple

import numpy as np
import pandas as pd

from .config import TraderConfig


def bybit_fee_fn(trade_value: float, config: TraderConfig) -> float:
    return trade_value * config.bybit_fee


def simulate_order_fill(
    direction: str,
    mid_price: float,
    config: TraderConfig,
    spread_bps: Optional[float] = None,
    slippage_bps: Optional[float] = None,
    reject_prob: Optional[float] = None,
) -> Tuple[Optional[float], str]:
    spread_bps = spread_bps if spread_bps is not None else config.spread_bps
    slippage_bps = slippage_bps if slippage_bps is not None else config.slippage_bps
    reject_prob = reject_prob if reject_prob is not None else config.order_reject_prob

    if random.random() < reject_prob:
        return None, "rejected"

    spread = mid_price * (spread_bps / 10000)
    slippage = abs(np.random.normal(slippage_bps, slippage_bps / 2))
    slippage_amt = mid_price * (slippage / 10000)

    if direction == "long":
        fill_price = mid_price + spread + slippage_amt
    else:
        fill_price = mid_price - spread - slippage_amt

    time.sleep(random.uniform(0, config.max_fill_latency))
    return fill_price, "filled"


def mark_to_market_equity(
    cash_equity: float,
    position: int,
    entry_price: Optional[float],
    qty: float,
    last_price: float,
    margin_used: float = 0.0,
) -> float:
    total = cash_equity + margin_used
    if position == 1 and entry_price:
        total += (last_price - entry_price) * qty
    if position == -1 and entry_price:
        total += (entry_price - last_price) * qty
    return total


def calc_liq_price_long(entry_price: float, leverage: int) -> float:
    return entry_price * (1 - 1 / leverage)


def calc_liq_price_short(entry_price: float, leverage: int, config: TraderConfig) -> float:
    """Approximate Bybit linear perp liquidation with maintenance margin + taker fee."""
    mm_rate = getattr(config, "maintenance_margin_rate", 0.004)
    taker_fee = config.bybit_fee
    liq_price = entry_price * (1 + (1 / leverage) - mm_rate + taker_fee)
    return max(liq_price, 0.0)


def allowed_leverage_for_notional(notional: float, tiers: Sequence[tuple[float, int]], max_leverage: int) -> int:
    """Return the maximum leverage Bybit would allow for the proposed notional."""
    sorted_tiers = sorted(tiers, key=lambda x: x[0])
    for threshold, lev in sorted_tiers:
        if notional <= threshold:
            return max(1, min(int(lev), max_leverage))
    # If above the largest threshold, fall back to the smallest leverage in the last tier
    return max(1, min(int(sorted_tiers[-1][1]), max_leverage))


def resolve_leverage(margin_used: float, desired_leverage: float, config: TraderConfig) -> float:
    """Clamp requested leverage against Bybit tier rules and the configured cap."""
    leverage = min(desired_leverage, config.bybit_max_leverage)
    if margin_used <= 0:
        return 1.0

    # Iteratively clamp leverage until notional fits within the allowed tier.
    while True:
        notional = margin_used * leverage
        allowed = allowed_leverage_for_notional(notional, config.bybit_leverage_tiers, config.bybit_max_leverage)
        if leverage <= allowed or leverage <= 1:
            return max(1.0, float(leverage))
        leverage = float(allowed)


@dataclass
class PositionState:
    side: Optional[str] = None  # "long" or "short"
    entry_price: Optional[float] = None
    tp_price: Optional[float] = None
    liq_price: Optional[float] = None
    qty: float = 0.0
    entry_bar_time: Optional[pd.Timestamp] = None
    entry_fee: float = 0.0
    trade_value: float = 0.0
    margin_used: float = 0.0
    leverage: float = 1.0
    exit_type: Optional[str] = None
    exit_target: Optional[float] = None

File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/Unity App Project/short_trader_multi_filter/Assets/Scripts/Core/OrderUtils.cs
using System;
using System.Collections.Generic;

namespace UnityApp.ShortTraderMultiFilter
{
    public static class OrderUtils
    {
        private static readonly Random Random = new();

        public static double BybitFee(double tradeValue, TraderConfig config) => tradeValue * config.BybitFee;

        public static (double? Price, string Status) SimulateOrderFill(
            string direction,
            double midPrice,
            TraderConfig config,
            int? spreadBps = null,
            int? slippageBps = null,
            double? rejectProbability = null)
        {
            var spread = (spreadBps ?? config.SpreadBps) / 10000.0;
            var slippage = (slippageBps ?? config.SlippageBps) / 10000.0;
            var rejectProb = rejectProbability ?? config.OrderRejectProbability;

            if (Random.NextDouble() < rejectProb)
            {
                return (null, "rejected");
            }

            var spreadAmount = midPrice * spread;
            var slippageAmount = midPrice * Math.Abs(NormalSample(slippage, Math.Max(slippage / 2.0, 0.00001)));

            var fillPrice = direction == "long"
                ? midPrice + spreadAmount + slippageAmount
                : midPrice - spreadAmount - slippageAmount;

            if (config.MaxFillLatencySeconds > 0)
            {
                var wait = Random.NextDouble() * config.MaxFillLatencySeconds;
                System.Threading.Thread.Sleep(TimeSpan.FromSeconds(wait));
            }

            return (fillPrice, "filled");
        }

        public static double MarkToMarket(
            double cashEquity,
            int positionSide,
            double? entryPrice,
            double quantity,
            double lastPrice,
            double marginUsed = 0.0)
        {
            var total = cashEquity + marginUsed;
            if (positionSide == 1 && entryPrice.HasValue)
            {
                total += (lastPrice - entryPrice.Value) * quantity;
            }
            else if (positionSide == -1 && entryPrice.HasValue)
            {
                total += (entryPrice.Value - lastPrice) * quantity;
            }
            return total;
        }

        public static double CalcLiquidationPriceLong(double entryPrice, int leverage) =>
            entryPrice * (1 - 1.0 / leverage);

        public static double CalcLiquidationPriceShort(double entryPrice, int leverage, TraderConfig config)
        {
            var maintenanceRate = 0.004;
            var takerFee = config.BybitFee;
            var liq = entryPrice * (1 + (1.0 / leverage) - maintenanceRate + takerFee);
            return Math.Max(liq, 0.0);
        }

        public static int AllowedLeverageForNotional(double notional, IReadOnlyList<(double Threshold, int Leverage)> tiers, int maxLeverage)
        {
            var sorted = new List<(double, int)>(tiers);
            sorted.Sort((a, b) => a.Threshold.CompareTo(b.Threshold));
            foreach (var (threshold, lev) in sorted)
            {
                if (notional <= threshold)
                {
                    return Math.Max(1, Math.Min(lev, maxLeverage));
                }
            }
            var last = sorted[^1].Leverage;
            return Math.Max(1, Math.Min(last, maxLeverage));
        }

        public static double ResolveLeverage(double marginUsed, double desiredLeverage, TraderConfig config)
        {
            var leverage = Math.Min(desiredLeverage, config.BybitMaxLeverage);
            if (marginUsed <= 0)
            {
                return 1.0;
            }

            while (true)
            {
                var notional = marginUsed * leverage;
                var allowed = AllowedLeverageForNotional(notional, config.BybitLeverageTiers, config.BybitMaxLeverage);
                if (leverage <= allowed || leverage <= 1)
                {
                    return Math.Max(1.0, leverage);
                }
                leverage = allowed;
            }
        }

        private static double NormalSample(double mean, double stdDev)
        {
            // Box-Muller transform
            var u1 = 1.0 - Random.NextDouble();
            var u2 = 1.0 - Random.NextDouble();
            var randStdNormal = Math.Sqrt(-2.0 * Math.Log(u1)) * Math.Sin(2.0 * Math.PI * u2);
            return mean + stdDev * randStdNormal;
        }
    }
}
